### PR TITLE
Add VolumeBindingMode to StorageClass API

### LINKED ISSUE #54434: Volume topology aware scheduling API changes
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind feature
/sig storage

Tracker bug for StorageClass API changes as specified in kubernetes/community#1168

/assign

### PR DESCRIPTION
**Pull request by msau42**
**What this PR does / why we need it**:
Adds a new field `VolumeBindingMode` to `StorageClass`, as specified in kubernetes/community#1168

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #54434

**Special notes for your reviewer**:
API changes only.  The scheduler and PV controller work will be submitted as a separate PR.

**Release note**:
NONE

@kubernetes/sig-storage-pr-reviews

### CODE REVIEW

File: pkg/apis/storage/types.go [hunk 1]
```diff
@@ -65,8 +65,19 @@ type StorageClass struct {
...
+	// provisioned and bound.  When unset, VolumeBindingImmediate is used.
+	// This field is alpha-level and is only honored by servers that enable
+	// the VolumeTopologyBinding feature.
+	// +optional
+	VolumeBindingMode *VolumeBindingMode
 }
 
+// NoProvisioner is a special Provisioner value to use when a StorageClass
+// is required but there is no dynamic provisioner.
+const NoProvisioner = "k8s.io/no-provisioner"
```
Comment (thockin) [inline pkg/apis/storage/types.go]: Should be `kubernetes.io/no-provisioner` - we don't generally use k8s.io in formal names (except API versions, but that one got away from me).

File: pkg/apis/storage/types.go [hunk 2]
```diff
@@ -80,3 +91,18 @@ type StorageClassList struct {
...
+const (
+	// VolumeBindingImmediate indicates that PersistentVolumeClaims should be
+	// immediately provisioned and bound.
+	VolumeBindingImmediate VolumeBindingMode = "Immediate"
+
+	// VolumeBindingTopologyAware indicates that PersistentVolumeClaims should not
+	// be provisioned and bound until a Pod is created that references the
+	// PeristentVolumeClaim.  The volume provisioning and binding will occur during
+	// Pod scheduing.
+	VolumeBindingTopologyAware VolumeBindingMode = "TopologyAware"
```
Comment (thockin) [inline pkg/apis/storage/types.go]: As a constant, this doesn't really explain what it will do.  What's wrong with "FirstConsumer" or something ?

Comment (msau42) [reply pkg/apis/storage/types.go]: "FirstPod"?  "WaitForPod"?

**Comment by thockin** [comment]
"First" seems like a valuable distinction.

File: pkg/apis/storage/types.go (hunk 2 above)
Comment (msau42) [reply pkg/apis/storage/types.go]: How about "WaitForFirstConsumer"?  I like how the "wait" contrasts with "immediate"

**Comment by thockin** [comment]
It's a mouthful, but we have worse.

**Comment by msau42** [comment]
Updated

File: pkg/apis/storage/types.go [hunk 3]
```diff
@@ -65,8 +65,19 @@ type StorageClass struct {
...
+	// provisioned and bound.  When unset, VolumeBindingImmediate is used.
+	// This field is alpha-level and is only honored by servers that enable
+	// the VolumeTopologyBinding feature.
+	// +optional
+	VolumeBindingMode *VolumeBindingMode
 }
 
+// NoProvisioner is a special Provisioner value to use when a StorageClass
+// is required but there is no dynamic provisioner.
+const NoProvisioner = "kubernetes.io/no-provisioner"
```
Comment (thockin) [inline pkg/apis/storage/types.go]: Why do we need this, as opposed to ""

also, this seems unrelated to this PR?

File: pkg/features/kube_features.go [hunk 4]
```diff
@@ -169,6 +169,12 @@ const (
 	//
 	// Enable nodes to exclude themselves from service load balancers
 	ServiceNodeExclusion utilfeature.Feature = "ServiceNodeExclusion"
+
+	// owner: @msau42
+	// alpha: v1.9
+	//
+	// Extend the default scheduler to be aware of PV topology and handle PV binding
+	VolumeTopologyBinding utilfeature.Feature = "VolumeTopologyBinding"
```
Comment (thockin) [inline pkg/features/kube_features.go]: This name feels awkward - it doesn't really capture what it is doing?

File: pkg/apis/storage/types.go (hunk 3 above)
Comment (msau42) [reply pkg/apis/storage/types.go]: Validation requires that the provisioner string is non-empty.  Also, when we add dynamic provisioning support in the future, we will need a way to distinguish between a StorageClass that only supports static provisioning (with some fake provisioner name) vs dynamic provisioners.  But you're right, we don't need it right now.  I can add it in a future PR when it will actually get used.

File: pkg/features/kube_features.go (hunk 4 above)
Comment (msau42) [reply pkg/features/kube_features.go]: Hmmm the end goal is to allow the scheduler to make pv binding decisions.  How about, VolumeScheduling?  Or VolumeBindingInScheduler?

**Comment by msau42** [comment]
Updated to call the feature "VolumeScheduling"

**Comment by jsafrane** [comment]
lgtm from my side

**Comment by msau42** [comment]
@thockin is the name more clear now?

File: pkg/apis/storage/v1/util/helpers.go [hunk 5]
```diff
@@ -51,3 +54,10 @@ func IsDefaultAnnotation(obj metav1.ObjectMeta) bool {
 
 	return false
 }
+
+// IsBindingModeWaitForFirstConsumer returns true if the VolumeBindingMode is set
+// to VolumeBindingWaitForFirstConsumer
+func IsBindingModeWaitForFirstConsumer(class *storagev1.StorageClass) bool {
```
Comment (thockin) [inline pkg/apis/storage/v1/util/helpers.go]: Under what conditions do we have a StorageClass with a nil mode?  Shouldn't that be handled in defaulting?  Ie why do we need this?

**Review by thockin — COMMENTED** [review]
I think you want to add this to pkg/apis/storage/v1/defaults.go so you can rely on non-nil

File: pkg/apis/storage/v1/util/helpers.go (hunk 5 above)
Comment (msau42) [reply pkg/apis/storage/v1/util/helpers.go]: I got the impression that it is not possible to default new alpha fields.  The process now requires that we drop all alpha fields before storing them in etcd if you don't have the feature gate enabled.

File: pkg/apis/storage/v1/defaults.go [hunk 6]
```diff
@@ -31,4 +33,9 @@ func SetDefaults_StorageClass(obj *storagev1.StorageClass) {
 		obj.ReclaimPolicy = new(v1.PersistentVolumeReclaimPolicy)
 		*obj.ReclaimPolicy = v1.PersistentVolumeReclaimDelete
 	}
+
+	if obj.VolumeBindingMode == nil && utilfeature.DefaultFeatureGate.Enabled(features.VolumeScheduling) {
```
Comment (msau42) [inline pkg/apis/storage/v1/defaults.go]: This doesn't work with upgrades.  Need to discuss with @kubernetes/sig-api-machinery-pr-reviews if it's possible to default new alpha fields.

Comment (liggitt) [reply pkg/apis/storage/v1/defaults.go]: When gate disabled, they should not be persisted or returned from the API, so… not really

Comment (msau42) [reply pkg/apis/storage/v1/defaults.go]: Discussed with @thockin and @lavalamp and verified that defaults are applied on read to objects that were stored with a new field.  So we can default new fields under a feature gate as long as we also access the fields under the feature gate.

Comment (liggitt) [reply pkg/apis/storage/v1/defaults.go]: Doesn't that mean the default values are serialized back out and returned to API clients even when the gate is disabled? If so, that isn't what we want

Comment (liggitt) [reply pkg/apis/storage/v1/defaults.go]: > So we can default new fields under a feature gate

Ah, you're saying the defaulting itself would be gated. I thought I looked into that at one point and it resulted in weird dependencies, but maybe I'm misremembering

**Comment by thockin** [comment]
Assign back to me once it is LGTM'ed and ready for approval.

**Comment by thockin** [comment]
Oops, canned reply :)

Assign to me when ready for a look, just trying to manage my queue

File: pkg/apis/storage/v1/defaults.go (hunk 6 above)
Comment (msau42) [reply pkg/apis/storage/v1/defaults.go]: Here's a scenario I'm not sure how to handle:

1.  1.9 feature disabled, create new object, nil field not defaulted
2.  1.10 feature is enabled, update object, defaulting will kick in.  How does validation detect the difference between nil field -> defaulting vs nil field -> user set some value.  I want this field to be immutable.

Comment (liggitt) [reply pkg/apis/storage/v1/defaults.go]: > How does validation detect the difference between nil field -> defaulting vs nil field -> user set some value

it doesn't (and shouldn't). the old object decoded from etcd will get the default, get sent to the user, and any update coming from the user without that value will also get the default. from validation's perspective, both the old and new object will have the default value, so the field will not appear to have changed.

**Comment by msau42** [comment]
Linter is failing due to reasons that were not caused by my change...  Do I need to add these packages to some ignore list?

defaults.go:31:1: exported function SetDefaults_StorageClass should have comment or be unexported
defaults.go:31:6: don't use underscores in Go names; func SetDefaults_StorageClass should be SetDefaultsStorageClass
doc.go:17:1: package comment should be of the form "Package v1 ..."
register.go:37:2: exported var AddToScheme should have comment or be unexported

**Comment by msau42** [comment]
Added defaulting and removed the helper function that was doing the defaulting previously.

/assign @thockin

**Comment by msau42** [comment]
Had to rebase, didn't change anything

**Comment by msau42** [comment]
This is ready to go @thockin
